### PR TITLE
end2end qa pipeline (#167)

### PR DESCRIPTION
**Pull request by rosequ**
- Pipeline for running end2end QA system
- Managed model dependencies through python packaging
- Added documentation and shared the additional required files.

### CODE REVIEW

**Comment by rosequ** [comment]
@nims11, @gauravbaruah please do a CR

File: docs/setup_qa_pipeline.md [hunk 1]
```diff
@@ -0,0 +1,42 @@
+# Steps to setup an end2end QA pipeline
```
Comment (salman1993) [inline docs/setup_qa_pipeline.md]: Can you please add this directory structure to the docs as well? Some may clone the directories in different places.

Get all the repos in castorini and arrange them in this format
- castorini
    -- Anserini
    -- Castor
    -- data
    -- models

File: docs/setup_qa_pipeline.md [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+Make sure that you have correctly set your `config` file. For help, refer 
+[this](https://github.com/castorini/Anserini/blob/master/docs/speech-ui-api-docs.md).
+
+Start the Gateway server, Flask API, and the javascript app:
+
+```
+chmod +x run_ui.sh
+./run_ui.sh
+```
```
Comment (salman1993) [inline docs/setup_qa_pipeline.md]: Mention that this won't work in Ubuntu:
- comment out the npm from the script 
- use curl or Postman to test the endpoint

File: src/main/python/api.py [hunk 3]
```diff
@@ -6,7 +6,13 @@
 from flask import Flask, jsonify, request
 # FIXME: separate this out to a classifier class where we can switch out the models
 from pyserini import Pyserini
-from jaccard import Jaccard
+from sm_model.bridge import SMModelBridge
+
+path_to_castorini = os.getcwd() + "/../../../.."
```
Comment (salman1993) [inline src/main/python/api.py]: This requires the directory structure to be maintained so it's important that they clone the repos in the same place.

Comment (rosequ) [inline src/main/python/api.py]: Yes, adding a typical structure in the doc should solve this problem...

**Comment by rosequ** [comment]
@salman1993 have another look and see if we can ship this...?

**Comment by salman1993** [comment]
ship it!

File: src/main/python/api.py (hunk 3 above)
Comment (tuzhucheng) [inline src/main/python/api.py]: This is highly fragile. It's ok for a temporary solution but I have an alternative proposal: #172

File: src/main/python/api.py [hunk 4]
```diff
@@ -6,7 +6,13 @@
 from flask import Flask, jsonify, request
 # FIXME: separate this out to a classifier class where we can switch out the models
 from pyserini import Pyserini
-from jaccard import Jaccard
+from sm_model.bridge import SMModelBridge
```
Comment (tuzhucheng) [inline src/main/python/api.py]: How does this work? I get ModuleNotFoundError: No module named 'sm_model'? sm_model is a part of Castor.

Comment (rosequ) [inline src/main/python/api.py]: @tuzhucheng did you follow these steps: https://github.com/castorini/Anserini/blob/master/docs/setup_qa_pipeline.md ?

File: docs/setup_qa_pipeline.md [hunk 5]
```diff
@@ -0,0 +1,57 @@
+# Steps to setup an end2end QA pipeline 
+Clone the Anserini, models, and data repo:
+```
+git clone https://github.com/rosequ/Anserini.git
```
Comment (nims11) [inline docs/setup_qa_pipeline.md]: Fix the url?

File: docs/setup_qa_pipeline.md [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+git clone https://github.com/castorini/data.git
+git clone https://github.com/castorini/models.git
```
Comment (nims11) [inline docs/setup_qa_pipeline.md]: Make these two installable by pip and add them to requirements.txt

Comment (rosequ) [inline docs/setup_qa_pipeline.md]: @nims11 only Anserini has python code and needs to be pip installed. @tuzhucheng clone the original Anserini repo and check if the issue repeats. I'll have the docs updated.

Comment (nims11) [inline docs/setup_qa_pipeline.md]: You don't need to have python code to make it a python package. We are just using pip and requirements.txt here as a module management tool.

**Comment by nims11** [comment]
I also suggest to enforce `virtualenv` usage through the docs. `requirements.txt` has static version number (which is how it should be) defined and many people like to keep their global python environment updated. `virtualenv` is highly recommended specially when multiple people are working on a python project.

**Comment by tuzhucheng** [comment]
I'm using `virtualenv` myself, need to document it sometime soon :)

**Comment by rosequ** [comment]
I use it too but left the details the users. We, should probably mention it in docs. @tuzhucheng is your issue resolved?

**Comment by tuzhucheng** [comment]
Yes @rosequ, I missed the install from requirements.txt - can get it working now!

**Pull request merged**